src/main/java/org/sonatype/nexus/repository/composer/internal/ComposerHostedMetadataFacetImpl.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Sonatype Nexus (TM) Open Source Version
+ * Copyright (c) 2018-present Sonatype, Inc.
+ * All rights reserved. Includes the third-party code listed at http://links.sonatype.com/products/nexus/oss/attributions.
+ *
+ * This program and the accompanying materials are made available under the terms of the Eclipse Public License Version 1.0,
+ * which accompanies this distribution and is available at http://www.eclipse.org/legal/epl-v10.html.
+ *
+ * Sonatype Nexus (TM) Professional Version is available from Sonatype, Inc. "Sonatype" and "Sonatype Nexus" are trademarks
+ * of Sonatype, Inc. Apache Maven is a trademark of the Apache Software Foundation. M2eclipse is a trademark of the
+ * Eclipse Foundation. All other trademarks are the property of their respective owners.
+ */
+package org.sonatype.nexus.repository.composer.internal;
+
+import java.io.IOException;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+
+import org.sonatype.nexus.common.event.EventManager;
+import org.sonatype.nexus.common.stateguard.Guarded;
+import org.sonatype.nexus.repository.FacetSupport;
+import org.sonatype.nexus.repository.storage.Asset;
+import org.sonatype.nexus.repository.storage.AssetCreatedEvent;
+import org.sonatype.nexus.repository.storage.AssetDeletedEvent;
+import org.sonatype.nexus.repository.storage.AssetEvent;
+import org.sonatype.nexus.repository.storage.AssetUpdatedEvent;
+import org.sonatype.nexus.repository.storage.StorageFacet;
+import org.sonatype.nexus.transaction.UnitOfWork;
+
+import com.google.common.eventbus.AllowConcurrentEvents;
+import com.google.common.eventbus.Subscribe;
+import org.joda.time.DateTime;
+
+import static jline.internal.Preconditions.checkNotNull;
+import static org.joda.time.DateTime.now;
+import static org.sonatype.nexus.common.stateguard.StateGuardLifecycleSupport.State.STARTED;
+import static org.sonatype.nexus.repository.composer.internal.ComposerAttributes.P_PROJECT;
+import static org.sonatype.nexus.repository.composer.internal.ComposerAttributes.P_VENDOR;
+import static org.sonatype.nexus.repository.storage.AssetEntityAdapter.P_ASSET_KIND;
+
+/**
+ * Concrete implementation of {@code ComposerHostedMetadataFacet} using events. Essentially a stripped-down form of the
+ * mechanism behind the Yum createrepo implementation in our internal codebase.
+ */
+@Named
+public class ComposerHostedMetadataFacetImpl
+    extends FacetSupport
+    implements ComposerHostedMetadataFacet
+{
+  private final EventManager eventManager;
+
+  @Inject
+  public ComposerHostedMetadataFacetImpl(final EventManager eventManager)
+  {
+    this.eventManager = checkNotNull(eventManager);
+  }
+
+  @Subscribe
+  @Guarded(by = STARTED)
+  @AllowConcurrentEvents
+  public void on(final AssetDeletedEvent deleted) {
+    if (matchesRepository(deleted) && isEventRelevant(deleted)) {
+      invalidateMetadata(deleted);
```

Comment:
Is there a concern of multiple deletes/adds/updates happening here? For yum I think there is a configurable hold off period to try and avoid invalidation happening constantly

Reply:
> Is there a concern of multiple deletes/adds/updates happening here? For yum I think there is a configurable hold off period to try and avoid invalidation happening constantly

For the Yum implementation we do have a hold-off period that's implemented _within_ the `invalidateMetadata()` method. 

However, I'd like to avoid doing that here until we find out that we need to in order to avoid the additional complexity. Unlike something that's being shipped as part of the main NXRM codebase I think we can be more flexible in what we do until we find out someone is encountering a problem of that nature at scale, then resolve it.  

Reasons I'm less concerned include:

* The rebuild metadata operation here should be a lot less onerous in that it's scoped to a particular vendor/project pair , so I would expect the number of assets to load and process to be on the order of tens rather than hundreds or thousands.
* While I can see lots of new builds being pushed into a large Yum repo and triggering rebuilds, in this case (much as in my previous point) the impact is limited to a particular vendor/project. So I suspect we'll have fewer problems with that (and fewer back-to-back rebuilds) than we would globally for a Yum repository. 

On the other hand, if we turn out to need that optimization, I would definitely do something similar to what was done for the Yum implementation as a place to start.

Reply:
The time buffer in Yum _reduces_ the number of metadata generations when multiple RPMs are being uploaded (i.e. if you upload 60 RPMs 1 a second for a minute you get 1 regeneration rather than 60), however there is another mechanism buried in that logic that avoid firing any more events than we need to. There was a problem whereby if the invalidation event was fired too many times that we depleted the thread pool in the event manager. That was likely more visible in yum _because_ we were waiting. So it's something to be aware of here but not necessarily change.

Given that the metadata is vendor and project scoped and this will be quicker I think it would take some serious effort to cause any issues (You'd likely need to upload 1000s of packages a minute).

As a side note we had a conversation recently about making it clearer what features are implemented/missing from this kind of work and also what kind of scale it is tested up to.